### PR TITLE
Rewrite KVS in dataflow

### PR DESCRIPTION
**Pull request by justinj**
This is a first pass at a dataflow based KVS. It is missing a bunch of clock
stuff, but I was getting anxious about not merging this so here's an early
version of it.

The structure of the program is roughly as follows:

```mermaid
graph BT
    Y[Read Requests] --> Z[Streaming Join]
    X[Received Batches] --> Z
    Z --> Q[Respond to Read Request]

    F[Write Requests] --> B
    A[Epoch Timer] --> B[Batcher / Merger]
    B --> C[Join]
    G[Ownership] --> C
    C --> E[Network Shuffle]
```

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
(Submitted an issue for rustfmt giving up: https://github.com/rust-lang/rustfmt/issues/5249 )

File: hydroflow/src/builder/build/mod.rs [hunk 1]
```diff
@@ -10,6 +10,7 @@ pub mod pull_handoff;
 pub mod pull_iter;
 pub mod pull_join;
 pub mod pull_map;
+pub mod pull_stream_join;
```
Comment (jhellerstein) [inline hydroflow/src/builder/build/mod.rs:13]: StreamJoin is an ambiguous name. This thing is asymmetric: one side is fully streaming/transient, the other side accumulates over time. I'd call it something like OneSidedHashJoin (as opposed to SymmetricHashJoin) or  StreamBufferHashJoin (in the old streaming database papers I think it'd be called a Stream-Table Join).

File: hydroflow/src/compiled/pull.rs [hunk 2]
```diff
@@ -11,6 +11,82 @@ where
...
+    tab: HashMap<K, L::Repr>,
+}
+
+impl<K, L> Default for StreamJoinState<K, L>
+where
+    L: LatticeRepr,
+{
+    fn default() -> Self {
+        Self {
+            tab: HashMap::new(),
```
Comment (jhellerstein) [inline hydroflow/src/compiled/pull.rs:27]: I'm kind of curious why the state doesn't have buffers as in SymmetricHashJoin. I would have expected this to look like "half of a SymmetricHashJoin".

File: hydroflow/src/compiled/pull.rs [hunk 3]
```diff
@@ -11,6 +11,82 @@ where
...
+    L: LatticeRepr + Merge<Update>,
+    L::Repr: Default,
+    Update: LatticeRepr,
+    StreamV: Clone,
+    Buf: Iterator<Item = (K, Update::Repr)>,
+    Stream: Iterator<Item = (K, StreamV)>,
+{
+    type Item = (K, StreamV, L::Repr);
+
+    fn next(&mut self) -> Option<Self::Item> {
```
Comment (jhellerstein) [inline hydroflow/src/compiled/pull.rs:59]: Again kind of curious why this isn't more explicitly similar to SymmetricHashJoin

File: hydroflow/src/compiled/pull.rs (hunk 2 above)
Comment (jhellerstein) [inline hydroflow/src/compiled/pull.rs:27]: I'm kind of curious why the state doesn't have buffers as in SymmetricHashJoin. I would have expected this to look like "half of a SymmetricHashJoin".

File: hydroflow/src/builder/build/mod.rs (hunk 1 above)
Comment (jhellerstein) [inline hydroflow/src/builder/build/mod.rs:13]: StreamJoin is an ambiguous name. This thing is asymmetric: one side is fully streaming/transient, the other side accumulates over time. I'd call it something like OneSidedHashJoin (as opposed to SymmetricHashJoin) or StreamBufferHashJoin (in the old streaming database papers I think it'd be called a Stream-Table Join).

File: kvs/src/main.rs [hunk 4]
```diff
@@ -1,35 +1,36 @@
 #![feature(never_type)]
```
Comment (jhellerstein) [inline kvs/src/main.rs:1]: Not gonna try to comment constructively here since I don't know this code well but I'm surprised to see so many changes! Should this be in a separate PR?

File: kvs/src/main.rs [hunk 5]
```diff
@@ -1,35 +1,36 @@
...
-        },
+        sync::mpsc::{channel, Receiver, Sender},
     },
 };
 
 #[derive(Debug, Copy, Clone, Hash, Eq, PartialEq)]
 struct ActorId(u64);
 
-type Timestamp = HashMap<usize, u64>;
+// type Timestamp = HashMap<usize, u64>;
```
Comment (jhellerstein) [inline kvs/src/main.rs:26]: Leave a comment to explain why this is commented out for now, and a reminder to come back and clean this up.

**Review by jhellerstein — COMMENTED** [review]
Don't have a lot of intelligent thoughts on changes to `builder`. My understanding of `main.rs` isn't very robust but I wonder if extensive changes there don't merit a separate PR.

File: hydroflow/src/builder/build/mod.rs (hunk 1 above)
Comment (MingweiSamuel) [reply hydroflow/src/builder/build/mod.rs:13]: Maybe HalfJoin or HalfHashJoin?

File: hydroflow/src/compiled/pull.rs (hunk 2 above)
Comment (MingweiSamuel) [reply hydroflow/src/compiled/pull.rs:27]: The buffer in SymmetricHashJoin is needed because an incoming value might result in multiple output values, but we can only output one value at a time. Instead of buffering, here we output all values together as a single lattice point, I think.

File: kvs/src/main.rs [hunk 6]
```diff
@@ -120,80 +141,123 @@ where
...
+                    hf.add_channel_input::<_, Option<Message<K, V>>, VecHandoff<Message<K, V>>>("writes");
+                // Feed incoming messages into a Hydroflow channel.
+                // TODO(justin): not sure why the tokio Receiver doesn't support try_iter?
+                tokio::spawn(async move {
+                    // TODO(justin): batch all these per try_recv?
+                    while let Some(v) = receiver.recv().await {
+                        q_send.give(Some(v));
+                        q_send.flush();
+                    }
+                });
```
Comment (MingweiSamuel) [inline kvs/src/main.rs:178]: You can use [`ReceiverStream`](https://docs.rs/tokio-stream/0.1.8/tokio_stream/wrappers/struct.ReceiverStream.html) to turn `receiver` into an async stream and use `add_input_from_stream` instead

**Review by MingweiSamuel — APPROVED** [review]
